File: src/main/java/net/database/FriendAPI.java
package net.database;

import client.Character;
import client.player.friend.FriendList;
import net.server.Server;
import org.jooq.Record;
import org.jooq.Result;

import static database.jooq.Tables.FRIENDS;

public class FriendAPI {

    public static void loadFriends(Character character) {
        Result<Record> friends = DatabaseCore.getConnection()
                .select().from(FRIENDS)
                .where(FRIENDS.CID.eq(character.getId()))
                .fetch();

        FriendList friendList = character.getFriendList();
        for (Record record : friends) {
            int fid = record.getValue(FRIENDS.FID);
            friendList.addFriend(fid, CharacterAPI.getOfflineName(fid), record.getValue(FRIENDS.GROUP));

            Character chr = Server.getInstance().getCharacter(fid);
            if (chr != null) {
                friendList.getFriends().get(fid).setChannel(chr.getChannel().getChannelId());
            }
        }
        friendList.updateFriendList();
        friendList.notifyMutualFriends();
    }

    public static void loadPending(Character character) {
        Result<Record> pending = DatabaseCore.getConnection()
                .select().from(FRIENDS)
                .where(FRIENDS.FID.eq(character.getId()))
                .and(FRIENDS.PENDING.eq((byte) 1))
                .fetch();

        FriendList friendList = character.getFriendList();
        for (Record record : pending) {
            int toPend = record.getValue(FRIENDS.CID);
            if (!friendList.getFriends().containsKey(toPend)) {
                friendList.getPending().add(toPend);
            }
        }
    }

    public static void addFriend(int cid, int fid, String group, boolean pending) {
        DatabaseCore.getConnection()
                .insertInto(FRIENDS,
                        FRIENDS.CID,
                        FRIENDS.FID,
                        FRIENDS.GROUP,
                        FRIENDS.PENDING)
                .values(cid, fid, group, (byte) (pending ? 1 : 0))
                .execute();
    }

    public static void removePendingStatus(int cid, int fid) {
        DatabaseCore.getConnection()
                .update(FRIENDS)
                .set(FRIENDS.PENDING, (byte) 0)
                .where(FRIENDS.CID.eq(cid))
                .and(FRIENDS.FID.eq(fid))
                .execute();
    }

    public static void removeFriend(int cid, int fid) {
        DatabaseCore.getConnection()
                .deleteFrom(FRIENDS)
                .where(FRIENDS.CID.eq(cid))
                .and(FRIENDS.FID.eq(fid))
                .execute();
    }

    public static void updateGroup(int cid, int fid, String groupName) {
        DatabaseCore.getConnection()
                .update(FRIENDS)
                .set(FRIENDS.GROUP, groupName)
                .where(FRIENDS.CID.eq(cid))
                .and(FRIENDS.FID.eq(fid))
                .execute();
    }
}

File: src/main/java/net/database/FriendAPI.kt
package net.database

import client.Character
import database.jooq.Tables
import database.jooq.Tables.FRIENDS
import net.database.DatabaseCore.connection
import net.server.Server

object FriendAPI {

    fun loadFriends(character: Character) {
        with(FRIENDS) {
            val friends = connection.select().from(this)
                .where(CID.eq(character.id))
                .fetch()
            val friendList = character.friendList

            friends.forEach {
                val fid = it.getValue(FID)
                val friend =
                    friendList.addFriend(fid, CharacterAPI.getOfflineName(fid), it.getValue(GROUP))

                val chr = Server.getCharacter(fid) ?: return@forEach
                friend.channel = chr.getChannel().channelId
            }
            friendList.updateFriendList()
            friendList.notifyMutualFriends()
        }
    }

    fun loadPending(character: Character) {
        with(FRIENDS) {
            val pending = connection.select().from(this)
                .where(FID.eq(character.id))
                .and(PENDING.eq(1.toByte()))
                .fetch()
            val friendList = character.friendList

            pending.forEach {
                val toPend = it.getValue(CID)

                if (!friendList.friends.containsKey(toPend)) {
                    friendList.pending.add(toPend)
                }
            }
        }
    }

    fun addFriend(cid: Int, fid: Int, group: String, pending: Boolean) {
        with(FRIENDS) {
            connection.insertInto(
                this,
                CID,
                FID,
                GROUP,
                PENDING
            ).values(cid, fid, group, (if (pending) 1 else 0).toByte()).execute()
        }
    }

    fun removePendingStatus(cid: Int, fid: Int) {
        with(FRIENDS) {
            connection.update(this)
                .set(PENDING, 0.toByte())
                .where(CID.eq(cid))
                .and(FID.eq(fid))
                .execute()
        }
    }

    fun removeFriend(cid: Int, fid: Int) {
        with(FRIENDS) {
            connection.deleteFrom(this)
                .where(CID.eq(cid))
                .and(FID.eq(fid))
                .execute()
        }
    }

    fun updateGroup(cid: Int, fid: Int, groupName: String) {
        with(FRIENDS) {
            connection.update(this)
                .set(GROUP, groupName)
                .where(CID.eq(cid))
                .and(FID.eq(fid))
                .execute()
        }
    }
}